sklearn/__init__.py

```diff
@@ -141,6 +143,10 @@
     except ModuleNotFoundError:
         pass
 
+    # Set a global controller that can be used to locally limit the number of
+    # threads without looping through all shared libraries every time.
+    _sklearn_threadpool_controller = ThreadpoolController()
```

Comment:
I added a comment. 

The BLAS libs are already loaded in a kind of hidden way. Importing utils will import both numpy and scipy.linalg at some point. The OpenMP lib is already loaded by importing show_versions. I now import numpy and scipy.linalg to make sure that the BLAS libs are loaded first in a more explicit way.

I wouldn't add a test because if threadpoolctl fails to discover the libs (like with the latest scipy), the test would fail without anything that can be done on sklearn's side.

Reply:
> I wouldn't add a test because if threadpoolctl fails to discover the libs (like with the latest scipy), the test would fail without anything that can be done on sklearn's side.

Fair enough, I guess this is quite unlikely that people will move this code around anyway. I guess you mean in Numpy 2 rc e.g. https://github.com/joblib/threadpoolctl/pull/175.

Reply:
Just curious BTW, if you create `_sklearn_threadpool_controller` too early (before loading OpenBLAS for example) the `_sklearn_threadpool_controller.limit(limits=1, user_api="blas")` don't do anything and there is no warning about it, right?

Reply:
> I guess you mean in Numpy 2

indeed :)

Reply:
> Just curious BTW, if you create _sklearn_threadpool_controller too early (before loading OpenBLAS for example) the _sklearn_threadpool_controller.limit(limits=1, user_api="blas") don't do anything and there is no warning about it, right?

Yes, the goal of the controller object it to look for loaded libs once, during instantiation. Then calls to ``limit`` won't refresh the list so it will have no effect if libs were not loaded before instantiation.

It would be hard to find a valid and consistent condition to warn about. For instance OpenMP is not mandatory and threadpoolctl can fail to detect BLAS libs (cf numpy 2) (it's a bug but still it can happen), so it can happen that no libs at all are loaded in the controller and we can't even warn in that situation.